### ISSUE TITLE
The click event is ignored when using an app indicator tray icon

### ISSUE THREAD

**Issue by seppestas** [electron/electron#6773]
As documented in the [Tray API docs](https://github.com/electron/electron/blob/master/docs/api/tray.md#event-click), the click event is ignored for tray icons that have been created as an app indicator, i.e on Ubuntu Unity.
This means ["menu bar"](https://github.com/maxogden/menubar) style applications do not work on Ubuntu (see maxogden/menubar#1), making them not really cross-platform.

The reason there is no click event being exposed is that these kind of applications do not fit the ["Ubuntu status menu design guidelines"](https://wiki.ubuntu.com/CustomStatusMenuDesignGuidelines) and therefor [libappindicator](https://bugs.launchpad.net/ubuntu/+source/libappindicator) does not support it.

The proper solution would be to have Ubuntu alter there style guides and make the app indicator adhere the [Status Notifier specification](https://www.freedesktop.org/wiki/Specifications/StatusNotifierItem/StatusNotifierItem/), or at least the `Activate` method. This has already been requested in [Unity libappindicator bug 522152](https://bugs.launchpad.net/ubuntu/+source/libappindicator/+bug/522152), which hasn't received much attention since it's creation in 2010, regardless of it's high bug-heat score.
Tl;dr: I don't think this will happen any time soon.

My plan is to work around the limitations set by libappindicator by listening directly to the signals sent by the Unity Panel service as [described](https://bugs.launchpad.net/ubuntu/+source/libappindicator/+bug/522152/comments/11) by @sjakthol. I already created a [PoC](https://github.com/seppestas/test-libappindicator).

**Comment by seppestas** [comment]
This issue will mostly require changes in Chromium. Would it be best to create an issue in [libchromiumcontent](https://github.com/electron/libchromiumcontent) and submit patches there, or to try and push changes directly to Chromium itself?

**Comment by seppestas** [comment]
Some related issues:
- gillesdemey/Cumulus#64
- Maybulb/Nimble#18
- jingweno/hacker-menu#18
- willyb321/electron-menubar-timetable#28

**Comment by willyb321** [comment]
ooh ive been referenced. Yeah, I can confirm that my application doesn't work on Ubuntu.

**Comment by zcbenz** [comment]
Considering the patch would be relatively large, I think pushing changes to Chromium would be a better choice, and it would benefit much more people than Electron users.

**Comment by seppestas** [comment]
I don't know if the changes would be useful outside of Electron, AFAIK Chrome/Chromium only uses the app indicator in a "Unity compatible" way. But I agree, the changes are pretty big, so trying to push them straight into the Chromium project might make more sense than making a huge patch file.

The problem I have with this is that this greatly increases the scope and impact of the changes, and I have no experience with the whole testing / [try server stuff](https://www.chromium.org/developers/testing/try-server-usage) (and I don't even know if I can use it). Currently, I just build using libchromiumcontent's build system and test manually, and even that is pushing the limits of my workstation (especially since I do it in a VM). I'm reading up on the Chromium project docs, but any pointers on how this can be done more smoothly are welcome.

PS. I don't really think this issue is a bug, IMO the `enhancement` label fits better. Also, since it only impacts distributions using app indicator (i.e Ubuntu Unity), `linux-distribution-specific` looks like a more specific label than `linux` (IMO it should even be called `distribution-specific`. There's nothing stopping you from running Gnome or Unity on top of e.g BSD). I'm planning to make a different issue for the GTK status icon.

**Comment by njaard** [comment]
I can confirm that electron tray icons don't appear to get click events on KWin/KDE.

**Comment by akashnimare** [comment]
@seppestas  My little menubar [app](https://github.com/akashnimare/foco) broke down because of this bug 😭 . Does anyone know how to fix it?

**Comment by seppestas** [comment]
I have been working towards a solution (but have been a bit busy with work and other projects recently), but most changes need to happen inside of chromium. For GTK status icons the changes are pretty small and sensible, but Ubuntu Unity's app indicator requires some more "hacky" workarounds.

I managed to get these workarounds working inside of chromium but I'm now looking into moving as much as possible to electron instead. The part I'm struggeling with now is figuring out what version of libappindicator is used by Chromium inside Electron.

**Comment by nerijus** [comment]
The  tray docs say "When app indicator is used on Linux, the click event is ignored". A few questions:
1. If I don't use libappindicator, tray left click will work?
2. How do I not use libappindicator with Electron on Fedora? Should I recompile electron without libappindicator support?
BTW, when I have my app running, lsof does not show libappindicator.so* or libappindicator3.so* opened by it.

**Comment by seppestas** [comment]
@nerijus, libappindicator is only used on Ubuntu Unity afaik (but that doesn't mean it can't be used on other systems).

I guess on your system a GTK status icon is used instead. The click event for GTK status icons works, but it does not contain the location of the icon which is needed if you want your application to be shown on a location relative to the status icon.

**Comment by nerijus** [comment]
I want my application to display its main window when I click on tray.
I ran the sample from issue 1715, it should print 'Tray clicked' on clicked event, but it does not.
I have libappindicator and libappindicator-gtk3 packages installed, but I renamed the .so* files before running the test.

**Comment by seppestas** [comment]
Maybe you could try running my test code at https://github.com/seppestas/test-libappindicator ?

If `test-unity-panel-service` and `test-menu-about-to-show-signal` work I would assume Electron will use libappindicator. If not, `test-gtk-status-icon` should work.

**Comment by nerijus** [comment]
I had to install libappindicator-devel and libdbusmenu-gtk2-devel to compile.
test-unity-panel-service launches tray icon, when I left or right click it shows some empty line.
test-menu-about-to-show-signal the same.
test-gtk-status-icon does not display tray icon.

**Comment by nerijus** [comment]
I rebuilt electron without libappindicator-devel package installed, but left click from #1715 still does not work.

**Comment by nerijus** [comment]
Changed tray.on('clicked)' to tray.on('click') in the sample from #1715, and left click works!

**Comment by pbaesse** [comment]
Any news about?? :)

**Comment by nerijus** [comment]
Please see my previous comment - it works on Fedora. As for Ubuntu, I don't know, someone has to try tray.on('click') in the sample from #1715.

**Comment by seppestas** [comment]
Nope, does not work on Ubuntu, by (Canonical) design.

The "official" workaround is registering a menu item the user can click. The problem is that menubar somewhat breaks this.

I have a working workaround (that only works on the first click), but I implemented it in Chromium instead of in Electron. I'm pretty sure I can get it to work by putting the code in electron itself though, the hardest part is figuring out the address of the Unity bar thingy.

Anyhow, I was hoping to get this fixed sooner, but I'm a bit busy with some other stuff atm.

**Comment by bbales** [comment]
@seppestas has there been any progress on this?

**Comment by seppestas** [comment]
I have PoC for my workaround implemented in Chromium, but after some discussion with @eglaysher I decided against trying include the code in the Chromium project. Since the workaround relies on undocumented internal signals of the Unity panel, the workaround could stop working when Ubuntu makes changes to their Panel code. My workaround will also break menu scrubbing in the app indicator area, to quote @eglaysher:

> IIRC, EntryActivated isn't what you want for this; it's explicitly not a click event. So, let's say you're a user and you have three appindicators in your bar, something system provided like the network menu, the system provided bluetooth menu and your electron app provided appindicator. The user mouse downs (and then mouse ups) on the network icon which displasy a menu. The user mouse moves from the network icon to under the bluetooth icon, which hides the first network menu and displays the bluetooth menu. The user continues to move their mouse to the electron app's icon (and not touch any buttons!), dismissing the first menu, but then breaking the menu scrubbing as your app treats the mouse over as a click, and you make your own x11 window which takes focus.
>
> I'm trying to steer you away from treating the appindicator area as a set of buttons because the appindicator model, both to the programmer and the user, is that of a menu, and you're going to break user expectations in not-so-subtle ways if you treat it as a set of buttons. The unity appindicator area is implemented with GtkMenu (and iirc, the kde one is implemented with QMenu), and uses dbusmenu to copy the GtkMenu/QMenu structure from the user program to the menu displayed by the appindicator widget, and then ferries menu events from the appindicator widget back to the GtkMenu/QMenu instance in your program.
> 
> The way you're proposing to do things will break menu scrubbing, including accidental scrubbing, which happens all the time. This is why I'm trying to push you to use something like chrome's current replacement menu item workaround."

The root cause of this issue is Ubuntu Unity treating the "status icon area" as a menu instead of as "a set of buttons displayed as icons", making Unity in essence incompatible with "menubar" style apps. Unless Canonical changes this, a proper fix for this issue is impossible.

I am now looking in to moving the workaround to Electron. In theory it is possible to move everything into the Electron source tree, but having the workaround as part of Chromium makes sense because:
- The workaround needs to know the name of the libappindicator library and [Chromium already has code that figures this out](https://chromium.googlesource.com/chromium/src/+/master/chrome/browser/ui/libgtkui/app_indicator_icon.cc#104).
- There is already [code that deals with low-level Unity functions in Chromium](https://chromium.googlesource.com/chromium/src/+/master/chrome/browser/ui/libgtkui/unity_service.cc), so my workaround fits way nicer in Chromium than in Electron.

Other options are moving the workaround to [Brightray](https://github.com/electron/brightray), but I'm not sure if this fits in the scope of Brightray, or adding it as a [patch to libchromiumcontent](https://github.com/electron/libchromiumcontent/tree/master/patches), but it's quite a lot of code to put in a patch...

What do you think is best option @zcbenz?

**Comment by zcbenz** [comment]
If a proper fix is not possible then I think we should not include the workaround in Electron, the best solution is probably to write a native module for app indicator.

**Comment by seppestas** [comment]
What do you mean? Write an external lib and add it as a vendor submodule? If so, to what? Electron or Brightray / libchromiumcontent?

**Comment by pronebird** [comment]
@seppestas I think he meant a node.js native module. That's what we do to polish rough edges of Electron/Chromium already..

**Comment by seppestas** [comment]
That should be feasable... But are there any reasons not to make this workaround part of Brightray?

**Comment by midenok** [comment]
[Slack](https://slack.com/) is affected under KDE: left click doesn't work, red notification freezes app.

**Comment by codebytere** [comment]
Given that we generally try to keep the Electron native API layer thin, and that there are ways like those posted above that can be used to accomplish the goal, this enhancement is better kept in Electron Userland. As @zcbenz said, the best solution is likely a node.js native module.

**Comment by davicorreiajr** [comment]
Any news? I'm developing a menu bar app as well https://github.com/davicorreiajr/spotify-now-playing, but I don't know how to better adapt to linux

**Comment by kashaiahyah85** [comment]
> Given that we generally try to keep the Electron native API layer thin, and that there are ways like those posted above that can be used to accomplish the goal, this enhancement is better kept in Electron Userland. As @zcbenz said, the best solution is likely a node.js native module.

Nobody has come up with a method to accomplish the goal. Care to comment again today? Or does the Linux developer simply abandon Electron if they want to make things work reliably? MacOS seems to be the only truly supported platform, Windows when possible and Linux is a happy coincidence at best.

**Comment by seppestas** [comment]
It was impossible to get this working with Unity since they were explicitly against exposing the click event to applications / did not support "menu-bar" style applications explicitly. The good news for this issue is that Ubuntu ditched Unity in favor of Gnome!

On GTK/Gnome, the GTK mouse click event is exposed, getting this integrated in Electron shouldn't be to difficult. See https://github.com/seppestas/test-libappindicator/blob/master/test-gtk-status-icon.c
I'm quite sure something similar exists for KDE/plasma.

The bad news for this issue is that I'm not using Electron / Linux a while ago :X I'm sure someone else should be able to figure this out.

**Comment by seppestas** [comment]
Had a quick look at the code, looks like some lines are missing here: https://github.com/electron/electron/blob/1c99a9b425dd026b4be939486b48bd48022fee82/shell/browser/ui/tray_icon_gtk.cc#L61

Should be trivial to fix. Does anyone know if KDE uses the same GTK tray icon when using electron?

**Comment by lashkevi** [comment]
Still unsolved in five years. It worked for some time and then stopped working again, so that I need to keep obsolete versions of Walz and Caprine on my computers.
